### ISSUE TITLE
shared webserver instance

### ISSUE THREAD

**Issue by tablatronix** [tzapu/WiFiManager#461]
I am testing this already, but I am using a getter and a shared_ptr, not sure which way will be final.
Ill post my branch later, I was adding some callbacks and testing what kind of parameters and callouts we would want to for templating and sharing webserver control.

There are several options, make server public, add getter with shared_pointer, and af ew others.
I am unsure which is the best design pattern for this, or how we want to control ownership.

Thoughts welcome.

**Comment by ageurtse** [comment]
Is there a way you could provide an example on how to run a shared webserver.

**Comment by tablatronix** [comment]
Sure, when I figure out how

**Comment by tablatronix** [comment]
dnsserver? dns server resolves domains

**Comment by spilz87** [comment]
Hi I see 
As you can see I'm not a specialist ^^
You understand the idea : captive portal point to a "/captiveportal" page not to the "/" default page

**Comment by spilz87** [comment]
An other idea could be to check if the client is connected to AP or from network , and link to different pages, if it is connected on AP the page as only few button which link to other specific pages, so the difference is only on the first page 
Do you see what I mean ? 
Maybe it's already what you do

**Comment by tablatronix** [comment]
Yeah the idea is to share the webserver instance so you can access in your own code, you can do this now by just making it public, but I have not decided if that is the final route. It might be better to makde it a shared pointer so that memory gets cleaned up better.

**Comment by spilz87** [comment]
A shared pointer allow you to keep the same webserver but the configuration of this webserver (.on(..) , etc.) is erase when captive portal is open, isn't it ? 
Do I understand well or I'm wrong ?
Let me know if it's not the good place to have this discussion

**Comment by tablatronix** [comment]
I wouldn't think so, not sure what you mean. webportal is always on when configportal is on

**Comment by spilz87** [comment]
Does the setupConfigPortal erase the setup of the server done in the setup() of the sketch ?

**Comment by tablatronix** [comment]
Ah I see what you are saying, yes that resets the webserver. hmm you have to setup yours after that, wm would have to be setup to allow that to be skipped, it also gets torn down on closeconfigportal

**Comment by tablatronix** [comment]
Curious , Did you add a new constructor to pass the webserver in , or use a setter?

I guess you could check by client ip, or get the clients attached to ap and check them for that ip request.

**Comment by spilz87** [comment]
Not in WM, it was a test in a personal library 

Ok i will look for this function :)

**Comment by spilz87** [comment]
If you remember the name, let me know ^^
Thanks

**Comment by tablatronix** [comment]
https://github.com/esp8266/Arduino/issues/2681

**Comment by spilz87** [comment]
Ok thanks,
I just test it, it works well, as expected only for client connected to AP. So it's possible to compare IP to know if it connected to AP or through network, and send a different first web page

**Comment by spilz87** [comment]
I just test something, with my iPhone, I don't know if it's true with android and I didn't get captive portal on my computer, but the captive portal seems to request the "/hotspot-detect.html" 
Is it a standard ?...

**Comment by tablatronix** [comment]
for now I am testing this out

* moving server to public, keeping it unique pointer
* adding `setWebServerCallback` that fires after webserver is setup and before route events are setup ( this allows user to override routes eg. /info )

testing added to DEV example
